waku/waku_sync/protocol.nim

```diff
@@ -0,0 +1,311 @@
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import
+  std/options,
+  stew/results,
+  chronicles,
+  chronos,
+  metrics,
+  libp2p/protocols/protocol,
+  libp2p/stream/connection,
+  libp2p/crypto/crypto,
+  eth/p2p/discoveryv5/enr
+import
+  ../common/nimchronos,
+  ../common/enr,
+  ../waku_core,
+  ../waku_enr,
+  ../node/peer_manager/peer_manager,
+  ./raw_bindings,
+  ./common,
+  ./session,
+  ./storage_manager
+
+logScope:
+  topics = "waku sync"
+
+type WakuSync* = ref object of LPProtocol
+  storage: Storage # Negentropy protocol storage 
+  peerManager: PeerManager
+  maxFrameSize: int
+
+  syncInterval: timer.Duration # Time between each syncronisation attempt
+  relayJitter: Duration # Time delay until all messages are mostly received network wide
+  transferCallBack: Option[TransferCallback] # Callback for message transfer.
+
+  pruneCallBack: Option[PruneCallBack] # Callback with the result of the archive query
+  pruneStart: Timestamp # Last pruning start timestamp
+  pruneOffset: timer.Duration # Offset to prune a bit more than necessary. 
+
+  periodicSyncFut: Future[void]
+  periodicPruneFut: Future[void]
+
+proc storageSize*(self: WakuSync): int =
+  self.storage.len
+
+proc ingessMessage*(self: WakuSync, pubsubTopic: PubsubTopic, msg: WakuMessage) =
+  if msg.ephemeral:
+    return
+  #TODO: Do we need to check if this message has already been ingessed? 
+  # because what if messages is received via gossip and sync as well?
+  # Might 2 entries to be inserted into storage which is inefficient.
+  let msgHash: WakuMessageHash = computeMessageHash(pubsubTopic, msg)
+
+  trace "inserting message into storage ",
+    hash = msgHash.toHex(), timestamp = msg.timestamp
+
+  if self.storage.insert(msg.timestamp, msgHash).isErr():
+    debug "failed to insert message ", hash = msgHash.toHex()
+
+proc calculateRange(jitter: Duration, syncRange: Duration = 1.hours): (int64, int64) =
+  var now = getNowInNanosecondTime()
+
+  # Because of message jitter inherent to Relay protocol
+  now -= jitter.nanos
```

Comment:
Do we need to consider nano granularity?
Wondering if a different language sync implementation cannot support nanoSec granularity. 
Can we use second granularity instead?

Reply:
The precision is in nanoseconds but the actual accuracy is in seconds. This was the simplest thing to do.